File: code/richediting/src/main/java/org/fbme/ide/richediting/editor/RichApplicationProjection.java
package org.fbme.ide.richediting.editor;

import jetbrains.mps.project.Project;
import org.fbme.ide.iec61499.adapter.fbnetwork.MPSNetworkInstanceReference;
import org.fbme.ide.platform.editor.EditorProjectionController;
import org.fbme.ide.platform.editor.SimpleEditorProjection;
import org.fbme.lib.iec61499.instances.NetworkInstance;
import org.jdom.Element;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.mps.openapi.model.SNode;

import javax.swing.*;
import java.util.Objects;

public class RichApplicationProjection extends SimpleEditorProjection {

    private final NetworkInstance myInstance;
    private final Project myProject;

    public static final String PERSISTENCE_KEY = "instance-ref";

    public RichApplicationProjection(SNode node, EditorProjectionController controller, String name, String[] hints, NetworkInstance instance, Project project) {
        super(node, controller, name, hints);
        myInstance = instance;
        myProject = project;
    }

    public NetworkInstance getInstance() {
        return myInstance;
    }

    @NotNull
    @Override
    public JComponent createHeaderComponent() {
        return new NetworkInstanceHeaderComponentProvider(myNode, myInstance, myProject).getComponent();
    }

    @Nullable
    @Override
    public Object getData(@NotNull @NonNls String string) {
        if (Objects.equals(RichEditorDataKeys.APPLICATION.getName(), string)) {
            return myInstance.getRootInstance().getDeclaration();
        }
        if (Objects.equals(RichEditorDataKeys.NETWORK_INSTANCE.getName(), string)) {
            return myInstance;
        }
        return null;
    }

    @Override
    public void saveState(@NotNull Element element) {
        element.setAttribute(PERSISTENCE_KEY, MPSNetworkInstanceReference.create(myInstance).serialize());
    }
}

File: code/richediting/src/main/kotlin/org/fbme/ide/richediting/editor/RichApplicationProjection.kt
package org.fbme.ide.richediting.editor

import jetbrains.mps.project.Project
import org.fbme.ide.iec61499.fbnetwork.MPSNetworkInstanceReference
import org.fbme.ide.platform.editor.EditorProjectionController
import org.fbme.ide.platform.editor.NavigatableDeclaration
import org.fbme.ide.platform.editor.SimpleEditorProjection
import org.fbme.lib.iec61499.declarations.ResourceDeclaration
import org.fbme.lib.iec61499.instances.NetworkInstance
import org.jdom.Element
import org.jetbrains.annotations.NonNls
import org.jetbrains.mps.openapi.model.SNode
import javax.swing.JComponent

class RichApplicationProjection(
        node: SNode,
        controller: EditorProjectionController,
        name: String,
        hints: Array<String>,
        val instance: NetworkInstance,
        project: Project
) : SimpleEditorProjection(node, controller, project, name, hints) {

    override fun getData(@NonNls string: String): Any? = when (string) {
        RichEditorDataKeys.APPLICATION.name -> instance.rootInstance.declaration
        RichEditorDataKeys.NETWORK_INSTANCE.name -> instance
        else -> null
    }

    override fun saveState(element: Element) {
        element.setAttribute(PERSISTENCE_KEY, MPSNetworkInstanceReference.create(instance).serialize())
    }

    override val subElements: List<NavigatableDeclaration>
        get() = NetworkInstanceNavigationSupport.navigatablePath(project, instance)
    override val rootElement: NavigatableDeclaration
        get() = NavigatableDeclaration.build(project, node, "Overview", "Overview")

    companion object {
        const val PERSISTENCE_KEY = "instance-ref"
    }
}
